Canister holding tank

DeeJayhX opened the pull request:
# Pull Request Template

### Purpose
_Describe the problem the PR fixes or the feature it introduces_<br>
_Don't forget to use "Fixes #issuenumber" to select issues and auto close them on merge_
Fixes #2297 

### Notes:
_Please enter any other relevant information here_
-Updated Canister GUI to reflect ability to refill tanks
-Added ability to click on canister with tank to insert the tank
--Will not allow if a tank is already inserted
-Added ability to eject tank from canister
--Will do nothing if no tank is inserted
-Added CanisterFillable item trait
--Should only be applied to items that can fit into and attach to a canister.
### Please make sure you have followed the self checks below before submitting a PR:

- Code is sufficiently commented
- Code is indented with tabs and not spaces
- The PR does not include any unnecessary .meta, .prefab or <b>.unity (scene) changes</b>
- The PR does not bring up any new compile errors
- The PR has been tested in editor
- Any new files are named using PascalCase (to avoid issues on case sensitive file systems)
- Any new / changed components follow the [Component Development Checklist](https://github.com/unitystation/unitystation/wiki/Component-Development-Checklist)
- Any new objects / items follow the [Creating Items and Objects Guide](https://github.com/unitystation/unitystation/wiki/Creating-Items-and-Objects%3A-Now-With-Prefab-Variants) (especially concerning the use of prefab variants)
- The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)
- The PR has been tested with round restarts.
- The PR has been tested on moving / rotating / rotated-before-joining matrices (if applicable)

DeeJayhX committed:
Added tank refilling from canisters

-Updated Canister GUI to reflect ability to refill tanks
-Added ability to click on canister with tank to insert the tank
--Will not allow if a tank is already inserted
-Added ability to eject tank from canister
--Will do nothing if no tank is inserted
-Added CanisterFillable item trait
--Should only be applied to items that can fit into and attach to a canister.

DeeJayhX committed:
Merge branch 'develop' into canister-holding-tank

DeeJayhX committed:
Fixed client side messages for Canister and it's GUI

Client side messages broken by rebase

fomalsd commented on UnityProject/Assets/Resources/ScriptableObjects/Traits/CommonTraitsSingleton.asset:
```diff
@@ -34,4 +35,4 @@ MonoBehaviour:
   Cable: {fileID: 11400000, guid: e39c048ec864b47a495a40f9c76af764, type: 2}
   Welder: {fileID: 11400000, guid: aaed69546e98a47d8b488a046d00ebe7, type: 2}
-  Shovel: {fileID: 11400000, guid: 35ecd7ae92ffc9f4d9e62613c2ad8cf8, type: 2}
+  Shovel: {fileID: 11400000, guid: b336a4abd9d45924cb48b49a49fdea2b, type: 2}
```
is shovel reference change intentional?

DeeJayhX replied:
To my knowledge, I didn't change it, Unity did. So I would classify it as Unintentional, but possibly correct?

DeeJayhX replied:
Reverted and pushed.

fomalsd commented on UnityProject/Assets/Scripts/UI/Canister/GUI_Canister.cs:
```diff
@@ -326,16 +360,108 @@ public void ServerEditReleasePressure(string newValue)
 		ControlTabs.CloseTab(Type, Provider);
 	}
+
+	public void EjectExternalTank()
```
Looks like this code is executed on clientside rather than serverside. NetTab UI logic is supposed to run on serverside and pass information back to clients using NetElements.
`PlayerManager.LocalPlayer` won't work on a headless server since there's no local player. Also there can be multiple tab 'peepers' at any given time, so keep tat in mind. They can be accessed from `Peepers` hashset, but it's highly advisable to display all tab related warnings within a tab (updating some label) instead of resorting to chat notifications for UX reasons.
Since you don't know who exactly initiated the ejection, just make ejected container appear on top of canister.

DeeJayhX replied:
Fixed.

fomalsd replied:
`Chat.AddExamineMsgToClient()` only works on clientside

DeeJayhX replied:
Fixed.

fomalsd commented on UnityProject/Assets/Scripts/Util/SweetExtensions.cs:
```diff
@@ -26,7 +26,7 @@ public static ItemAttributesV2 Item(this GameObject go)
 	{
 		var item = go.Item();
-		if (item != null && !String.IsNullOrWhiteSpace(item.ArticleName)) return item.ArticleName;
+		if (item != null && !String.IsNullOrWhiteSpace(item.ArticleName)) return item.ArticleName.Replace("(Clone)","");
```
hmm, I think this is more of a bandaid for a deeper issue. ArticleName isn't supposed to have `(Clone)` in its name in the first place. maybe you could keep it as is for the time being

DeeJayhX replied:
@fomalsd 
> hmm, I think this is more of a bandaid for a deeper issue. ArticleName isn't supposed to have (Clone) in its name in the first place. maybe you could keep it as is for the time being

The code is as follows before the fix:
```csharp
public static string ExpensiveName(this GameObject go)
{
    var item = go.Item();
    if (item != null && !String.IsNullOrWhiteSpace(item.ArticleName)) return item.ArticleName

    var player = go.Player();
    if (player != null && !String.IsNullOrWhiteSpace(player.Name)) return player.Name;

    return go.name.Replace("NPC_", "").Replace("_", " ").Replace("(Clone)","");
}
```

Any GasContainer is considered an item.

The way I'm displaying the name is:
`externalTankName.SetValue = insertedContainer.ExpensiveName();`

which is displaying (Clone) at the end, because anything that returns as an item doesn't filter out (Clone).

I see a few solutions to this:
-Add a replace for (Clone) for items, which was my original solution
-Revert the replacement for (Clone) on items and use `netLabel.SetValue = item.Item().InitialName;`
instead of `netLabel.SetValue = item.ExpensiveName();`

Would the latter be preferred? Seems logical to me, and works in testing, but I'm curious if there's an issue I'm missing with it.

DeeJayhX replied:
Fixed:
Reverted the replacement for (Clone) on items in SweetExtensions.cs and use `netLabel.SetValue = item.Item().InitialName;`
instead of `netLabel.SetValue = item.ExpensiveName();`

fomalsd reviewed (changes requested):
Thanks for contribution!
UI's looking good, but ejection won't work on a real server due to it being clientside while it should've been serverside. Wrote suggestions below.

DeeJayhX committed:
Reverted changes to ExpensiveName and fixed tooltip

-Reverted changes to ExpensiveName filtering out (Clone) for items.
--Updated GUI_Canister.cs to utilize `Item().InitialName` instead of `ExpensiveName()`
-Fixed a tooltip in GasCanister being applied to the wrong attribute

DeeJayhX committed:
Updated item spawn and messages from local client to serverside

-Canister.cs
--updated EjectInsertedContainer to utilize serverside itemdrop on top of canister
--updated messages also not require variables from offline implementation
-GUI_Canister.cs
--updated call to EjectInsertedCanister() to not include player parameter, since it was local

DeeJayhX committed:
Reverted shovel commontrait reference

Reverted shovel reference in CommonTraitsSingleton.asset as it was modified by unity.

fomalsd commented on UnityProject/Assets/Scripts/UI/Canister/GUI_Canister.cs:
```diff
@@ -326,16 +360,107 @@ public void ServerEditReleasePressure(string newValue)
+		container.Opened = isOpen;
+		if (isOpen)
+		{
+			Chat.AddExamineMsgToClient($"Canister releasing at {container.ReleasePressure}");
```
here, too

DeeJayhX replied:
Fixed.

fomalsd commented on UnityProject/Assets/Scripts/UI/Canister/GUI_Canister.cs:
```diff
@@ -326,16 +360,107 @@ public void ServerEditReleasePressure(string newValue)
+		}
+		else if (usingTank && externalTank == null)
+		{
+			Chat.AddExamineMsgToClient("Tank valve lockout occured! You should close the valve, " +
```
here as well. note how `AddLocalMsgToChat` was used here before

DeeJayhX replied:
Fixed.

fomalsd reviewed (changes requested):
Something is wrong: canister pressure never drops (or just never gets refreshed) while filling tank certainly gets filled with gas. See the gif.
A pressure counter for inserted tank would be really useful as well!
![GIF](https://user-images.githubusercontent.com/10403536/74646579-0cbc2580-5172-11ea-863e-3eac6dbeea9a.gif)
Also that inserted tank icon is barely visible, perhaps you could just use a generic red tank icon for that, like one you showed in early screenshots?

DeeJayhX committed:
Added Internal Pressure Update on Tankfill and External Tank Pressure Dial

-Added Internal Pressure Update on Tankfill
--Main canister pressure will now update to reflect new pressure after refilling an external tank
-External Tank Pressure Dial
--Displays pressure of tank on insert

DeeJayhX committed:
Fixed canister displays and gas exchange

-Fixed messages to display inside the canister GUI
-Added showing correct pressure for external tank
-Set proper mole values for oxygentank and gascontainer prefabs

DeeJayhX committed:
Update TabCanister.prefab - fix levers

Levers were "sliding" off-center when turned. Fixed that.

fomalsd reviewed (approved):
sweet, thatks!
